File: 560-subarray-sum-equals-k/subarray-sum-equals-k.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function(nums, k) {
    let d={0:1}
    let preSum=0
    let count=0
    for(let i=0;i<nums.length;i++){
        preSum=preSum+nums[i]
        target=preSum-k
        if(target in d){
            count=count+d[target]
        }
        if(preSum in d){
            d[preSum]=d[preSum]+1
        }else{
            d[preSum]=1
        }
    }
    return count;
    
};

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {

        int count=0;
        int prefixSum=0;
        HashMap<Integer,Integer> countMap=new HashMap<>();
        countMap.put(0,1);

        for(int i=0;i<nums.length;i++){
            prefixSum+=nums[i];
            int target=prefixSum-k;
            if(countMap.containsKey(target)){
                count=count+countMap.get(target);
            }
            countMap.put(prefixSum,countMap.getOrDefault(prefixSum,0)+1);
        }

        return count;
        
    }
}